moveit_ros/occupancy_map_monitor/src/occupancy_map_monitor.cpp

```diff
@@ -39,124 +39,95 @@
 #include <moveit_msgs/srv/load_map.hpp>
 #include <moveit/occupancy_map_monitor/occupancy_map.h>
 #include <moveit/occupancy_map_monitor/occupancy_map_monitor.h>
+#include <moveit/occupancy_map_monitor/occupancy_map_monitor_middleware_handle.hpp>
 #include <boost/bind.hpp>
 
 namespace occupancy_map_monitor
 {
-static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor");
-
-OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node, double map_resolution)
-  : map_resolution_(map_resolution), debug_info_(false), mesh_handle_count_(0), node_(node), active_(false)
+namespace
 {
-  initialize();
+static const rclcpp::Logger LOGGER = rclcpp::get_logger("moveit.ros.occupancy_map_monitor");
 }
 
-OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node,
+OccupancyMapMonitor::OccupancyMapMonitor(std::unique_ptr<MiddlewareHandle> middleware_handle,
                                          const std::shared_ptr<tf2_ros::Buffer>& tf_buffer,
                                          const std::string& map_frame, double map_resolution)
-  : tf_buffer_(tf_buffer)
+  : middleware_handle_(std::move(middleware_handle))
+  , tf_buffer_(tf_buffer)
   , map_frame_(map_frame)
   , map_resolution_(map_resolution)
   , debug_info_(false)
   , mesh_handle_count_(0)
-  , node_(node)
+  , active_(false)
 {
   initialize();
 }
 
+OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node, double map_resolution)
+  : OccupancyMapMonitor(std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node), nullptr, "", map_resolution)
+{
+}
+
+OccupancyMapMonitor::OccupancyMapMonitor(const rclcpp::Node::SharedPtr& node,
+                                         const std::shared_ptr<tf2_ros::Buffer>& tf_buffer,
+                                         const std::string& map_frame, double map_resolution)
+  : OccupancyMapMonitor(std::make_unique<OccupancyMapMonitorMiddlewareHandle>(node), tf_buffer, map_frame,
+                        map_resolution)
+{
+}
+
 void OccupancyMapMonitor::initialize()
```

Comment:
I guess it would break the API and is therefore a no-go but I've been starting to lean on removing initialization functions since it is the constructor whose job it is to make sure the object is in usable state.

Reply:
In this case, initialize is a private method that is only called by the constructor.  I'm moving its code into the constructor and removing the method to make it more clear.  That change does not break API in this case.

Reply:
> but I've been starting to lean on removing initialization functions

:100: